src/metrics/metrics.ts

```diff
@@ -350,3 +357,138 @@ export function getMateAndAssists(pgnMoveLine: string) {
+    }
+  }
+
+  return {
+    gameLinkWithMostMoves,
+    maxNumMoves,
+  };
+
+}
+
+export async function getPieceLevelMoveInfo(games: FileReaderGame[]) {
+  const numMovesByPiece = {};
+  let averageNumMovesByPiece = {};
+  let pieceWithMostMovesInAGame = null;
+  let pieceWithHighestAverageNumMoves = null;
+  let gameLinkWithPieceMostMoves = null;
+  let numMovesMadePieceWithMostMoves = 0;
+  
+  let gameCount = 0;
+
+  for (const game of games) {
+    gameCount++;
+    const chess = new Chess();
+    const moveGenerator = chess.historyGenerator(game.moves);
+
+    const numMovesByPieceThisGame = {};
+
+    // update move counts of each unambiguous piece
+    for (let moveInfo of moveGenerator) {
+      const { move } = moveInfo;
+
+      let movedPiece = move.unambiguousSymbol;
+
+      if (movedPiece) {
+        if (!numMovesByPiece[movedPiece]){
+          numMovesByPiece[movedPiece] = 0
+        }
+        numMovesByPiece[movedPiece]++;
+        numMovesByPieceThisGame[movedPiece]++
+
+        // Check if the move is a castling move
+        if (moveInfo.move.flags === 'k' || moveInfo.move.flags === 'q') {
+          let movingRook;
+    
+          if (moveInfo.move.flags === 'k') {
+            movingRook = 'rh';
+          } else {
+            movingRook = 'ra';
+          }
+    
+          if (moveInfo.move.color === 'w') {
+            movingRook = movingRook.toUpperCase();
+          }
+
+          if (!numMovesByPiece[movingRook]){
+            numMovesByPiece[movingRook] = 0
+          }
+
+          // duplicative action for the array capturing moves by piece for this game
+          if (!numMovesByPieceThisGame[movingRook]){
+            numMovesByPieceThisGame[movingRook] = 0
+          }
+
+          numMovesByPiece[movingRook]++;
+          numMovesByPieceThisGame[movingRook]++;
+
+        }
+      }
+
+      // console.log(move.originalString)
+      // console.log(numMovesByPiece)
+
+    }
+
+    for (const uahPiece of Object.keys(numMovesByPieceThisGame)) {
+      let maxMovesInGame = numMovesByPieceThisGame[uahPiece];
+
+      if (maxMovesInGame > numMovesMadePieceWithMostMoves) {
+        numMovesMadePieceWithMostMoves = maxMovesInGame;
+        pieceWithMostMovesInAGame = uahPiece;
+        gameLinkWithPieceMostMoves = game.metadata
+        .find((item) => item.startsWith('[Site "'))
+        ?.replace('[Site "', '')
+        .replace('"]', '');
+      }
+    }
+
+  }
+
+  // calculate average num moves by piece
+  for (const uahPiece of Object.keys(numMovesByPiece)) {
+    averageNumMovesByPiece[uahPiece] = numMovesByPiece[uahPiece] / gameCount
+  }
+
+  let maxAverageNumMoves = 0;
+
+  // find the piece with the highest average num moves
+  for (const uahPiece of Object.keys(averageNumMovesByPiece)) {
+    const averageNumMoves = averageNumMovesByPiece[uahPiece];
+    if (averageNumMoves > maxAverageNumMoves) {
+      maxAverageNumMoves = averageNumMoves;
+      pieceWithHighestAverageNumMoves = uahPiece;
+    }
```

Comment:
think there is a `max` method that applies to collections

Reply:
more efficient than this? if yes worth trying, otherwise won't overdo for now

Reply:
readability, i.e.

const highestAverageMoves = max(averageNumMovesByPiece[uahPiece])